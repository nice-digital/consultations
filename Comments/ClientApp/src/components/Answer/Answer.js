// @flow

import React, { Component, Fragment } from "react";
import { TextAnswer } from "./../AnswerTypes/TextAnswer/TextAnswer";
import { YesNoAnswer } from "./../AnswerTypes/YesNoAnswer/YesNoAnswer";

type PropsType = {
	staticContext?: any,
	answer: AnswerType,
	readOnly: boolean,
	saveAnswerHandler: Function,
	deleteAnswerHandler: Function,
	unique: string,
	updateUnsavedIds: Function,
};

type StateType = {
	answer: AnswerType,
	unsavedChanges: boolean,
};

export class Answer extends Component<PropsType, StateType> {
	constructor() {
		super();
		this.state = {
			answer: {},
			unsavedChanges: false,
		};
	}

	componentDidMount() {
		this.setState({
			answer: this.props.answer,
		});
	}

	textareaChangeHandler = (e: SyntheticEvent) => {
		const answer = this.state.answer;
		answer.answerText = e.target.value;
		const unsavedChanges = !(answer.answerId === -1 && answer.answerText.length === 0);
		this.props.updateUnsavedIds(`${answer.questionId}q`, unsavedChanges);
		this.setState({
			answer,
			unsavedChanges,
		});
	};

	yesNoChangeHandler = (e: SyntheticEvent) => {
		const answer = this.state.answer;
		answer.answerBoolean = (e.target.value === "true");
		const unsavedChanges = !(answer.answerId === -1 && answer.answerBoolean === undefined);
		this.props.updateUnsavedIds(`${answer.questionId}q`, unsavedChanges);
		this.setState({
			answer,
			unsavedChanges,
		});
	};

	static getDerivedStateFromProps(nextProps: any, prevState: any) {
		const prevTimestamp = prevState.answer.lastModifiedDate;
		const nextTimestamp = nextProps.answer.lastModifiedDate;
		const hasAnswerBeenUpdated = () => prevTimestamp !== nextTimestamp;
		if (hasAnswerBeenUpdated()) {
			return {
				answer: nextProps.answer,
				unsavedChanges: false,
			};
		}
		return null;
	}

	render() {
		if (!this.state.answer) return null;
		const {
			answerBoolean,
			answerText,
			answerId,
			questionId,
			commenterEmail,
		} = this.state.answer;
		const unsavedChanges = this.state.unsavedChanges;
		const answer = this.state.answer;
		const readOnly = this.props.readOnly;
		const questionText = this.props.questionText;
		const unique = this.props.unique;
		return (

			<Fragment>
				<section role="form">
					<form onSubmit={e => this.props.saveAnswerHandler(e, answer, this.props.questionId)} className="mb--0 mt--e">

						{commenterEmail &&
							<p className="CommentBox__commentBy mb--0">Answer by: {commenterEmail}</p>
						}

						{this.props.questionType.type === "YesNo" &&
						<YesNoAnswer
							unique={unique}
							questionText={questionText}
							readOnly={readOnly}
							textareaChangeHandler={this.textareaChangeHandler}
							yesNoChangeHandler={this.yesNoChangeHandler}
							answerText={answerText}
							answerBoolean={answerBoolean}
						/>
						}

						{this.props.questionType.type === "Text" &&
						<TextAnswer
							unique={unique}
							questionText={questionText}
							readOnly={readOnly}
							textareaChangeHandler={this.textareaChangeHandler}
							answerText={answerText}
						/>
						}

						{!readOnly && ((answerText && answerText.length > 0) || (answerBoolean !== undefined)) ?
							unsavedChanges ?
								<input
									data-qa-sel="submit-button"
									className="btn ml--0 mb--0"
									type="submit"
									value="Save answer"
								/>
								:
								<span className="ml--0 mb--0 CommentBox__savedIndicator">Saved</span>
							:
							null
						}

						{!readOnly && answerId > 0 &&
						<button
							data-qa-sel="delete-comment-button"
							className="btn mr--0 mb--0 right"
							onClick={e => this.props.deleteAnswerHandler(e, questionId, answerId)}>
							Delete
						</button>
						}

					</form>
				</section>
			</Fragment>
		);
	}
}
