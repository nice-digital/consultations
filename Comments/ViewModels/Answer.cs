using System;

namespace Comments.ViewModels
{
	public class Answer
	{
		public Answer() { } //only here for model binding. don't use it in code.
		public Answer(int answerId, string answerText, bool answerBoolean, DateTime lastModifiedDate, string lastModifiedByUserId, int questionId, int statusId)
		{
			AnswerId = answerId;
			AnswerText = answerText;
			AnswerBoolean = answerBoolean;
			QuestionId = questionId;
			LastModifiedDate = lastModifiedDate;
			LastModifiedByUserId = lastModifiedByUserId;
			StatusId = statusId;
		}

		public Answer(Models.Answer answer, Models.Question question = null)
		{
			AnswerId = answer.AnswerId;
			AnswerText = answer.AnswerText;
			AnswerBoolean = answer.AnswerBoolean;
			QuestionId = answer.QuestionId;
			LastModifiedDate = answer.LastModifiedDate;
			LastModifiedByUserId = answer.LastModifiedByUserId;
			StatusId = answer.StatusId;
			if (answer.Status != null)
				Status = new Status(answer.Status);

			if (question != null)
				SourceURI = question.Location.SourceURI;
			else if (answer.Question?.Location != null)
				SourceURI = answer.Question.Location.SourceURI;

			if (answer.OrganisationUser != null)
				CommenterEmail = answer.OrganisationUser.EmailAddress;
		}

		public int AnswerId { get; set; }
		public string AnswerText { get; set; }
		public bool? AnswerBoolean { get; set; }

		public int QuestionId { get; set; }

		public DateTime LastModifiedDate { get; set; }
		public string LastModifiedByUserId { get; set; }

		public int StatusId { get; set; }
		public ViewModels.Status Status { get; set; }

		/// <summary>
		/// This property doesn't match the database. it's only in a denormalised capacity to make things easier for the front-end.
		/// </summary>
		public string SourceURI { get; set; }

		public string CommenterEmail { get; set; }

        public bool showWhenFiltered { get; set; } = true;

		public void UpdateStatusFromDBModel(Models.Status status)
		{
			StatusId = status.StatusId;
			Status = new Status(status);
		}
	}
}
