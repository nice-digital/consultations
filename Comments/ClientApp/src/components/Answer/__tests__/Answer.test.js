import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Answer } from "../Answer";
import answerWithAnswer from "./answerWithAnswer.json";
import answerWithoutAnswer from "./answerWithoutAnswer.json";

const answerPropsWithAnswer = {
	answer: answerWithAnswer,
	readOnly: false,
	saveHandler: jest.fn(),
	deleteHandler: jest.fn(),
	unique: "string",
	updateUnsavedIds: jest.fn(),
	questionType: {
		type: "Text",
	},
};

const answerPropsWithoutAnswer = {
	answer: answerWithoutAnswer,
	readOnly: false,
	saveHandler: jest.fn(),
	deleteHandler: jest.fn(),
	unique: "string",
	updateUnsavedIds: jest.fn(),
	questionType: {
		type: "Text",
	},
};

test("sets text area with comment text correctly", async () => {
	render(<Answer {...answerPropsWithAnswer} />);
	expect(screen.getByDisplayValue("some answer text")).toBeInTheDocument();
});

test("unsavedChanges function is fired correctly on text area change", async () => {
	render(<Answer {...answerPropsWithAnswer} />);
	const textArea = screen.getByDisplayValue("some answer text");
	const user = userEvent.setup();
	textArea.focus();
	await user.type(textArea, " that's been updated");
	user.tab();
	expect(screen.getByDisplayValue("some answer text that's been updated")).toBeInTheDocument();
	expect(answerPropsWithAnswer.updateUnsavedIds).toHaveBeenCalledWith("22q", true);
});

test("should match snapshot with answer", () => {
	const {container} = render(<Answer {...answerPropsWithAnswer} />);
	expect(container).toMatchSnapshot();
});

test("should match snapshot without answer", () => {
	const {container} = render(<Answer {...answerPropsWithoutAnswer} />);
	expect(container).toMatchSnapshot();
});
