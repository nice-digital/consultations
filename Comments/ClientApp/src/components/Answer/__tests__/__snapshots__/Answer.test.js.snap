// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot with answer 1`] = `
<div>
  <section
    role="form"
  >
    <form
      class="mb--0 mt--e"
    >
      <div
        class="form__group form__group--textarea mb--b"
      >
        <label
          class="form__label visually-hidden"
          for="string"
        />
        <textarea
          class="form__input form__input--textarea"
          data-hj-whitelist="true"
          data-qa-sel="Comment-text-area"
          id="string"
        >
          some answer text that's been updated
        </textarea>
      </div>
      <span
        class="ml--0 mb--0 CommentBox__savedIndicator"
      >
        Saved
      </span>
      <button
        class="btn mr--0 mb--0 right"
        data-qa-sel="delete-comment-button"
      >
        Delete
      </button>
    </form>
  </section>
</div>
`;

exports[`should match snapshot without answer 1`] = `
<div>
  <section
    role="form"
  >
    <form
      class="mb--0 mt--e"
    >
      <div
        class="form__group form__group--textarea mb--b"
      >
        <label
          class="form__label visually-hidden"
          for="string"
        />
        <textarea
          class="form__input form__input--textarea"
          data-hj-whitelist="true"
          data-qa-sel="Comment-text-area"
          id="string"
        />
      </div>
      <span
        class="ml--0 mb--0 CommentBox__savedIndicator"
      >
        Saved
      </span>
    </form>
  </section>
</div>
`;
